ffi/derive/src/convert.rs

```diff
@@ -0,0 +1,315 @@
+                #( #discriminants )*
+
+                match source {
+                    #( #discriminant_names => Ok(#enum_name::#variant_names), )*
+                    _ => Err(iroha_ffi::FfiReturn::TrapRepresentation),
+                }
+            }
+        }
+        impl<'itm> iroha_ffi::TryFromReprC<'itm> for &'itm #enum_name {
+            type Source = *const #ffi_type;
+            type Store = ();
+
+            unsafe fn try_from_repr_c(source: <Self as iroha_ffi::TryFromReprC<'itm>>::Source, _: &mut <Self as iroha_ffi::TryFromReprC<'itm>>::Store) -> iroha_ffi::Result<Self> {
+                #( #discriminants )*
+
+                unsafe { match *source {
+                    #( | #discriminant_names )* => Ok(&*(source as *const _ as *const _)),
+                    _ => Err(iroha_ffi::FfiReturn::TrapRepresentation),
+                }}
+            }
+        }
+        impl<'itm> iroha_ffi::TryFromReprC<'itm> for &'itm mut #enum_name {
+            type Source = *mut #ffi_type;
+            type Store = ();
+
+            unsafe fn try_from_repr_c(source: <Self as iroha_ffi::TryFromReprC<'itm>>::Source, _: &mut <Self as iroha_ffi::TryFromReprC<'itm>>::Store) -> iroha_ffi::Result<Self> {
+                #( #discriminants )*
+
+                unsafe { match *source {
+                    #( | #discriminant_names )* => Ok(&mut *(source as *mut _ as *mut _)),
+                    _ => Err(iroha_ffi::FfiReturn::TrapRepresentation),
+                }}
+            }
+        }
+
+        impl<'slice> iroha_ffi::slice::TryFromReprCSliceRef<'slice> for #enum_name {
+            type Source = iroha_ffi::slice::SliceRef<'slice, Self>;
+            type Store = ();
+
+            unsafe fn try_from_repr_c(source: <Self as iroha_ffi::slice::TryFromReprCSliceRef<'slice>>::Source, _: &mut <Self as iroha_ffi::slice::TryFromReprCSliceRef<'slice>>::Store) -> iroha_ffi::Result<&'slice [Self]> {
+                source.into_rust().ok_or(iroha_ffi::FfiReturn::ArgIsNull)
+            }
+        }
+        impl<'slice> iroha_ffi::slice::TryFromReprCSliceMut<'slice> for #enum_name {
+            type Source = iroha_ffi::slice::SliceMut<'slice, #enum_name>;
+            type Store = ();
+
+            unsafe fn try_from_repr_c(source: <Self as iroha_ffi::slice::TryFromReprCSliceMut<'slice>>::Source, _: &mut <Self as iroha_ffi::slice::TryFromReprCSliceMut>::Store) -> iroha_ffi::Result<&'slice mut [Self]> {
+                source.into_rust().ok_or(iroha_ffi::FfiReturn::ArgIsNull)
+            }
+        }
+    }
+}
+
+fn derive_into_ffi_for_opaque_item(name: &Ident) -> TokenStream2 {
+    #[cfg(not(feature = "client"))]
+    let owned_into_ffi = quote! {
+        impl iroha_ffi::IntoFfi for #name {
+            type Target = *mut Self;
+
+            fn into_ffi(self) -> Self::Target {
+                Box::into_raw(Box::new(self))
```

Comment:
Is `Box`-ing and then immediately unboxing the best or fastest way to implement it?

Reply:
In other words, should we set up a ticket to make this better?

Reply:
not sure what you mean. This is where ownership is transfered across the FFI. The structure needs to be placed on the heap to escape local function context.

Reply:
Unboxing seems to be rather cheap operation, all we are doing is leaking `Box` and obtaining resulting pointer.

Reply:
And if we needed more control where it's allocated, with e.g. a custom allocator?

Reply:
`allocator_api` is unstable for now.

Reply:
I'll paraphrase: if we do it via a direct allocation, using the `unsafe` keyword and requesting direct access to the allocator we 
1. Clearly communicate that this is where the programmer modifying the code needs to be careful
2. Have more freedom in where and how we want the memory to be allocated (we know that this slab can be reused via a custom allocator)
3. Are more direct about our intention. In more detail: 

`Box::new` is a move with allocation. `Box::into_raw_parts` consumes the box, but retains (leaks) the allocated memory. A user doesn't see `unsafe` assumes that it's trivial and assumes that memory isn't leaked, when in reality it is.

Reply:
better now?